### ISSUE TITLE
compute kMACs/pixel

### ISSUE THREAD

**Issue by chaohuang** [lyq133/LIC-HPCM#12]
Hi Yuqi,

Could you please shed some light on how the complexity of your models was computed in terms of kMACs per pixel?

Thank you.

**Comment by lyq133** [comment]
Hi, thank you for your interest.

We computed the model complexity using the DeepSpeed FLOPs profiler on a single forward pass. For reference, the script is as follows:

```python
import torch
import importlib
import deepspeed

hw = 256
x = torch.zeros([1, 3, hw, hw]).cuda()

model_list = ['HPCM_Base', 'HPCM_Large']
param_list = []
mac_list = []
kmac_per_pixel_list = []

for model_name in model_list:
    model = importlib.import_module(f'.{model_name}', 'src.models').HPCM().cuda()
    model.eval()

    with torch.no_grad():
        flops_all, macs_all, params_all = deepspeed.profiling.flops_profiler.profiler.get_model_profile(
            model=model,
            args=[x],
            kwargs={},
            print_profile=False,
            detailed=False,
            warm_up=10,
            as_string=False,
        )

    param_list.append(params_all)
    mac_list.append(macs_all)

    kmacs_per_pixel = macs_all / (hw * hw) / 1e3
    kmac_per_pixel_list.append(kmacs_per_pixel)

print(param_list)
print(mac_list)
print(kmac_per_pixel_list)
```

Hope this helps!

**Comment by lyq133** [comment]
> Hello! I'm also very interested in this work. Perhaps we can communicate and learn from each other! My email is [1981893349@qq.com](mailto:1981893349@qq.com).

You can feel free to contact [lyq010303@mail.ustc.edu.cn](mailto:lyq010303@mail.ustc.edu.cn).

**Comment by chaohuang** [comment]
Hi Yuqi,

Thanks for the code to compute the model complexity. It works as expected; I’m getting 918.56 kMACs/pixel for the base model and 1261.32 kMACs/pixel for the large model, which are very close to the values reported in your paper.

I'm now trying to reduce the complexity of the base model by decreasing the steps. As a first attempt, I reduced the steps in scale-3 from 6 to 3. With this change, the decoding complexity dropped to 663.89 kMACs/pixel, which  is again close to your reported results.

However, after training the modified base model for 750 steps with a batch size of 8, the performance is very poor—the PSNR on the Kodak dataset is below 10 dB.

Would you mind taking a look at the changes [here](https://github.com/chaohuang/LIC-HPCM/commit/c6903ce659ae55b078042af166b5362803a4adc0#diff-bbbe748688e6283f625d07ff573b1992c68daf41f188e04f1f80c7548932e4aa) (only a few lines) to see if they are correct?

Thanks again.

**Comment by lyq133** [comment]
> Hi Yuqi,
> 
> Thanks for the code to compute the model complexity. It works as expected; I’m getting 918.56 kMACs/pixel for the base model and 1261.32 kMACs/pixel for the large model, which are very close to the values reported in your paper.
> 
> I'm now trying to reduce the complexity of the base model by decreasing the steps. As a first attempt, I reduced the steps in scale-3 from 6 to 3. With this change, the decoding complexity dropped to 663.89 kMACs/pixel, which is again close to your reported results.
> 
> However, after training the modified base model for 750 steps with a batch size of 8, the performance is very poor—the PSNR on the Kodak dataset is below 10 dB.
> 
> Would you mind taking a look at the changes [here](https://github.com/chaohuang/LIC-HPCM/commit/c6903ce659ae55b078042af166b5362803a4adc0#diff-bbbe748688e6283f625d07ff573b1992c68daf41f188e04f1f80c7548932e4aa) (only a few lines) to see if they are correct?
> 
> Thanks again.

Hi, thanks for the detailed report and for sharing the modification.

Could you please provide more information about the reported behavior? For example:
- whether the training loss decreases normally,
- and whether the poor result appears already in the direct `forward()` reconstruction, or mainly in the `compress()` / `decompress()` path.

My initial guess is that the issue may come from a compression / decompression consistency problem rather than the forward path itself. Since your current scale-3 mask list is still defined based on `get_mask_eight_parts(...)`, please first check the compression part around lines 390–391 and the decompression part around lines 561 and 564. Alternatively, you may also modify the definition of the scale-3 mask list so that it is based on `get_mask_four_parts(...)` instead.

**Comment by chaohuang** [comment]
The training and testing logs are attached:
[train_steps233_lambda0.0483.log](https://github.com/user-attachments/files/27028160/train_steps233_lambda0.0483.log)
[test_kodak.log](https://github.com/user-attachments/files/27028176/test_kodak.log)

We can see that the training process seems normal. For testing, the encoding PSNRs computed on the encoder side look reasonable (encoding PSNR is computed by using `y_hat` from `forward_hpcm` to compute `x_hat`), only the PSNRs of decoded images are very low, but I don't see how the decoder changes don't match the encoder changes ...

**Comment by lyq133** [comment]
> The training and testing logs are attached: [train_steps233_lambda0.0483.log](https://github.com/user-attachments/files/27028160/train_steps233_lambda0.0483.log) [test_kodak.log](https://github.com/user-attachments/files/27028176/test_kodak.log)
> 
> We can see that the training process seems normal. For testing, the encoding PSNRs computed on the encoder side look reasonable (encoding PSNR is computed by using `y_hat` from `forward_hpcm` to compute `x_hat`), only the PSNRs of decoded images are very low, but I don't see how the decoder changes don't match the encoder changes ...

Thanks for the logs. I will try to train your modified model on my side and debug it. This may take a few days, and I will report back if I have a clearer conclusion.

**Comment by lyq133** [comment]
> The training and testing logs are attached: [train_steps233_lambda0.0483.log](https://github.com/user-attachments/files/27028160/train_steps233_lambda0.0483.log) [test_kodak.log](https://github.com/user-attachments/files/27028176/test_kodak.log)
> 
> We can see that the training process seems normal. For testing, the encoding PSNRs computed on the encoder side look reasonable (encoding PSNR is computed by using `y_hat` from `forward_hpcm` to compute `x_hat`), only the PSNRs of decoded images are very low, but I don't see how the decoder changes don't match the encoder changes ...

Hi, I have now checked the modified 3-step scale-3 version more carefully. The issue was indeed in the compress/decompress consistency rather than the forward reconstruction path.

In the 3-step scale-3 setting, two original 8-part masks are merged into one coding step. However, in the previous modification, the merged scale-3 masks were still packed with the old writing logic, which is not a valid inverse of the corresponding decompression logic. As a result, the encoder-side reconstruction from `forward_hpcm()` could look reasonable, but the symbols written into the bitstream could not be correctly restored during `decompress_hpcm()`, leading to very poor decoded PSNR.

I fixed this by keeping the paired scale-3 masks explicitly during entropy coding. Specifically, I added separate helper functions to pack and restore each pair of 8-part masks, and updated both the scale-3 compression and decompression paths to use the same paired-mask logic. With this change, the modified model can now perform normal compression and decompression.

I have uploaded the updated model file for reference: 

[HPCM_Base_233.py](https://github.com/user-attachments/files/27236342/HPCM_Base_233.py)

**Comment by chaohuang** [comment]
> > The training and testing logs are attached: [train_steps233_lambda0.0483.log](https://github.com/user-attachments/files/27028160/train_steps233_lambda0.0483.log) [test_kodak.log](https://github.com/user-attachments/files/27028176/test_kodak.log)
> > We can see that the training process seems normal. For testing, the encoding PSNRs computed on the encoder side look reasonable (encoding PSNR is computed by using `y_hat` from `forward_hpcm` to compute `x_hat`), only the PSNRs of decoded images are very low, but I don't see how the decoder changes don't match the encoder changes ...
> 
> Hi, I have now checked the modified 3-step scale-3 version more carefully. The issue was indeed in the compress/decompress consistency rather than the forward reconstruction path.
> 
> In the 3-step scale-3 setting, two original 8-part masks are merged into one coding step. However, in the previous modification, the merged scale-3 masks were still packed with the old writing logic, which is not a valid inverse of the corresponding decompression logic. As a result, the encoder-side reconstruction from `forward_hpcm()` could look reasonable, but the symbols written into the bitstream could not be correctly restored during `decompress_hpcm()`, leading to very poor decoded PSNR.
> 
> I fixed this by keeping the paired scale-3 masks explicitly during entropy coding. Specifically, I added separate helper functions to pack and restore each pair of 8-part masks, and updated both the scale-3 compression and decompression paths to use the same paired-mask logic. With this change, the modified model can now perform normal compression and decompression.
> 
> I have uploaded the updated model file for reference:
> 
> [HPCM_Base_233.py](https://github.com/user-attachments/files/27236342/HPCM_Base_233.py)

Great! It works. Thank you, Yuqi.
